EIPS/eip-BN256_HashToCurve.md

```diff
@@ -0,0 +1,342 @@
+    h = ECAdd(h0, h1)
+    g = ECMul(h, r)     # Clear cofactor; g is now an element of G2
+    return g
+end function
+```
+
+Here is the pseudocode for BaseToTwist; the only real difference
+between BaseToG1 and BaseToTwist are the different constants
+and the different finite fields.
+All of these operations are performed in the finite field
+of the twist curve.
+As in BaseToG1, `inverse`, `is_square`, and `sqrt` perform
+the same operations in the underlying finite field GF(p^2).
+Our algorithms for `is_square` and `sqrt` come from this
+[paper](https://eprint.iacr.org/2012/685).
+
+```
+function BaseToTwist(t)
+    # All operations are done in the finite field GF(p^2)
+    # Here, the twist curve satisfies the equation
+    #       y^2 == g'(x) == x^3 + b'
+    p1 = (−1 + sqrt(−3))/2
+    p2 = −3
+    p3 = 1/3
+    p4 = g'(1)
+    s = (p4 + t^2)^3
+    alpha = inverse(t^2*(p4 + t^2))
+    x1 = p1 − p2*t^4*alpha
+    x2 = 1 − x1
+    x3 = −1 − p3*s*alpha
+    a1 = x1^3 + b'
+    a2 = x2^3 + b'
+    r1 = is_square(a1)
+    r2 = is_square(a2)
+    i = (r1 − 1)*(r2 − 3)/4 + 1
+    c1 = CTEq(1, i)
+    c2 = CTEq(2, i)
+    c3 = CTEq(3, i)
+    x = c1*x1 + c2*x2 + c3*x3
+    y = sign0(t)*sqrt(x^3 + b')
+    return (x, y)
+end function
+```
+
+## Rationale
+The current design choices reflect one choice of hashing algorithm.
+There is freedom in choosing the HashToBase function
+and this could easily be changed.
+Within HashToBase, the particular hashing algorithm
+(Keccak256 in our case) could also be modified.
+It may be desired to change the call to `sign0`
+at the end with `is_square`, as this would result
+in the same deterministic map to curve from the
+Fouque and Tibouchi
+[paper](https://www.di.ens.fr/~fouque/pub/latincrypt12.pdf)
+and ensure HashToG1 is indifferentiable from a random oracle;
+they proved this result in their paper.
+It may be possible to show that switching the `is_square`
+call with `sign0` does not affect indifferentiability,
+although this has not been proven at this time.
+
+[EIP-2537](./eip-2537.md)
+would allow for similar BLS signatures to be computed on
+the BLS12-381 elliptic curve.
+This proposal calls for additional precompiled contracts
+to allow for full use of pairing-based cryptography with
+this elliptic curve.
+This is the same rationale for wanting to add a hash-to-curve
+algorithm for BN256.
+
+## Backwards Compatibility
+There are no backward compatibility concerns.
+
+## Test Cases
+No underlying algorithms are being changed and so there is no need
+to specify additional test cases.
+
+## Implementation
+An implementation based on go-ethereum is forthcoming.
+It will contain the aforementioned functions and the necessary
+additions to the go-ethereum/crypto/bn256/cloudflare library.
+These additions could also be ported to the other Ethereum platforms.
+The required operations are inverses, square roots,
+and Legendre symbols in GF(p) and GF(p^2).
+These algorithms are straight-forward to implement.
```

Comment:
```suggestion
TBD
```
In general, it is better just leave information out than include information that is *intended* to be deleted later.  By putting `TBD` in here it is less likely that it will get missed in final review if this section is never updated.

Also, keep in mind that the Implementation section isn't required for an EIP, so you can just leave the whole section out until you have one ready to be included.

Reply:
Implementation section has been reduced to TBD.